Set missing hive stats to unknown instead of zero

rschlussel2 opened the pull request:
Previously we weren't checking that thrift fields were set before using their
values.  As a result, unset fields were reported as having a value of
zero.  This fixes that issue to appropriately set those fields as unknown.

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java:
```diff
@@ -243,8 +243,8 @@ public static HiveColumnStatistics fromMetastoreApiColumnStatistics(ColumnStatis
                     OptionalLong.empty(),
-                    OptionalLong.of(longStatsData.getNumNulls()),
-                    OptionalLong.of(longStatsData.getNumDVs()));
+                    longStatsData.isSetNumNulls() ? OptionalLong.of(longStatsData.getNumNulls()) : OptionalLong.empty(),
```
This repeating pattern could probably be captured to a method:
```
private static OptionalLong valueIfSet(boolean isSet, long value);
```

rschlussel2 replied:
Skipping this. We do a lot of the same thing in this section of the code with other types too e.g. Optional<T> and OptionalDouble.  Extracting a method makes it harder to tell that they're all the same.

findepi replied:
i didn't quite understand the explanation, but that's not too important.

kokosing commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/TestThriftMetastoreUtil.java:
```diff
@@ -0,0 +1,237 @@
+        assertEquals(actual.getDistinctValuesCount(), OptionalLong.empty());
+    }
+
+}
```
no new line :)

kokosing commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/TestThriftMetastoreUtil.java:
```diff
@@ -0,0 +1,237 @@
+package com.facebook.presto.hive.metastore.thrift;
```
no copyright

kokosing commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/TestThriftMetastoreUtil.java:
```diff
@@ -0,0 +1,237 @@
+public class TestThriftMetastoreUtil
+{
+    @Test
+    public void testLongStatsToColumnStatistics()
```
can please you divide these tests to one with stats and one with empty stats?

kokosing commented:
Other than my comments above. Great find!

rschlussel2 commented:
Addressed comments.  Squashed everything together because it's a pretty small change anyway.

I don't think I can reproduce the problem in `TestHiveTableStatistics` because it only occurs if you have partition stats for a partition but it's missing some fields.  I think these fields are never missing if you run analyze table, and if the table isn't analyzed, you don't get here at all.

I did reproduce this with real tables that were missing ndvs (stats weren't collected through hive analyze table).

findepi commented:
>  it only occurs if you have partition stats for a partition but it's missing some fields.

Did you try analyzing table and then altering the table to add some new columns?

rschlussel2 commented:
@findepi clever idea, but it didn't work.  There are still no ColumnStatisticObjs for those added columns, so this function doesn't get called.
